cmd/k8s-operator/spdy-hijacker.go

```diff
@@ -0,0 +1,189 @@
+			h.log.Error(err)
+		}
+		return
+	}()
+	cl := tstime.DefaultClock{}
+	lc := &spdyRemoteConnRecorder{
+		log:  h.log,
+		Conn: conn,
+		lw: &loggingWriter{
+			start:           cl.Now(),
+			clock:           cl,
+			failOpen:        h.failOpen,
+			sessionRecorder: wc,
+			log:             h.log,
+		},
+	}
+
+	qp := h.req.URL.Query()
+	ch := CastHeader{
+		Version:     asciicastv2,
+		Timestamp:   lc.lw.start.Unix(),
+		ExecCommand: strings.Join(qp["command"], " "),
+		SrcNode:     strings.TrimSuffix(h.who.Node.Name, "."),
+		SrcNodeID:   h.who.Node.StableID,
+		Namespace:   h.ns,
+		Pod:         h.pod,
+	}
+	if !h.who.Node.IsTagged() {
+		ch.SrcNodeUser = h.who.UserProfile.LoginName
+		ch.SrcNodeUserID = h.who.Node.User
+	} else {
+		ch.SrcNodeTags = h.who.Node.Tags
+	}
+	lc.ch = ch
+	return lc, nil
+}
+
+// CastHeader is the asciicast header to be sent to the recorder at the start of
+// the recording of a session.
+// https://docs.asciinema.org/manual/asciicast/v2/#header
+type CastHeader struct {
+	// Version is the asciinema file format version.
+	Version int `json:"version"`
+
+	// Width is the terminal width in characters.
+	Width int `json:"width"`
+
+	// Height is the terminal height in characters.
+	Height int `json:"height"`
+
+	// Timestamp is the unix timestamp of when the recording started.
+	Timestamp int64 `json:"timestamp"`
+
+	// Tailscale-specific fields: SrcNode is the full MagicDNS name of the
+	// tailnet node originating the connection, without the trailing dot.
+	SrcNode string `json:"srcNode"`
+
+	// SrcNodeID is the node ID of the tailnet node originating the connection.
+	SrcNodeID tailcfg.StableNodeID `json:"srcNodeID"`
+
+	// SrcNodeTags is the list of tags on the node originating the connection (if any).
+	SrcNodeTags []string `json:"srcNodeTags,omitempty"`
+
+	// SrcNodeUserID is the user ID of the node originating the connection (if not tagged).
+	SrcNodeUserID tailcfg.UserID `json:"srcNodeUserID,omitempty"` // if not tagged
+
+	// SrcNodeUser is the LoginName of the node originating the connection (if not tagged).
+	SrcNodeUser string `json:"srcNodeUser,omitempty"`
+
+	// Kubernetes-specific fields:
+	// Namespace of the Pod that is being exec-ed to.
+	Namespace string `json:"namespace,omitempty"`
+	// Name of the Pod that is being exec-ed to.
+	Pod string `json:"pod,omitempty"`
+	// ExecCommand is the command passed to 'kubectl exec' i.e 'sh' in 'kubectl exec -it my-pod sh'.
+	// Note that a Command field should not be used to store this info as
+	// that will make tsrecorder consider that the session consists of a
+	// non-interactive command.
+	ExecCommand string
+}
+
+func closeConnWithWarning(conn net.Conn, msg string) error {
+	b := io.NopCloser(bytes.NewBuffer([]byte(msg)))
+	resp := http.Response{Status: http.StatusText(http.StatusForbidden), StatusCode: http.StatusForbidden, Body: b}
+	if err := resp.Write(conn); err != nil {
+		return multierr.New(err, conn.Close())
+	}
+	return conn.Close()
```

Comment:
nit: this can be simplified

```suggestion
	return multierr.New(resp.Write(conn), conn.Close())
```

Reply:
That to me feels like it's harder to follow the flow/be certain that we will definitely first write and then close?

Reply:
That's fair 👍